### ISSUE TITLE
Test coverage not working for patters destrueturing of `dynamic` typed fields whem omiting type decalration

### ISSUE THREAD

**Issue by renancaraujo** [dart-lang/coverage#457]
Right now, it is impossible to attain 100% test coverage if you use pattern destructuring in your codebase on structures typed as `dynamic`. 

### To reproduce

`main.dart`
```dart
String thisDoesStuff(Map<String, dynamic> map) {
  final {'hello': world, 'count': count} = map;
  return 'Hello $world, count: $count';
}
```

`main_test.dart`
```dart
void main() {
  test('stuff', () {
    var myJson = <String, dynamic>{
      'hello': 'world',
      'count': [1, 2, 3],
    };
    expect(thisDoesStuff(myJson), 'Hello world, count: [1, 2, 3]');
  });
}
```

In those examples, generated coverages will mark `final {'hello': world, 'count': count} = map;` as an uncovered line.

If we change that line to `var {'hello': dynamic world, 'count': dynamic count} = map;`, it will **mark** as covered.

**Comment by liamappelbe** [comment]
Interesting. I've seen a few reports of a [tuple destructuring bug](https://github.com/flutter/flutter/issues/132663), which was fixed already (not by me though). But it seems the map destructuring case was missed, because I can confirm that this is still reproducible at head. Should be an easy fix though.

**Comment by liamappelbe** [comment]
It looks like the fix for tuple destructing was accidental. Bisecting found this commit: https://github.com/dart-lang/sdk/commit/4efcda0fd21643519486b041be87a56341f8ece7

**Comment by liamappelbe** [comment]
@johnniwinther I could use some help with this bug. There have been a bunch of bugs related to coverage for `PatternVariableDeclaration`s. For a while, record/tuple destructuring expressions were being incorrectly marked as not hit, but that was inadvertently fixed by this commit of yours: https://github.com/dart-lang/sdk/commit/4efcda0fd21643519486b041be87a56341f8ece7. However,map destructuring expressions still have this issue. I've been poking at `ConstantsTransformer.visitPatternVariableDeclaration` to figure out what's going on:

```Dart
// Map destructuring:
final {'hello': world, 'count': count} = map;

// AST before transformation:
PatternVariableDeclaration(final {"hello": var world, "count": var count} = map;)

// AST after transformation:
Block({
  final dynamic world;
  final dynamic count;
  {
    final Map<String, dynamic> #0#0 = map;
    late final dynamic #0#3 = #0#0.{Map.[]}("hello");
    late final dynamic #0#6 = #0#0.{Map.[]}("count");
    if (!((!(#0#3 == null) || null is dynamic && #0#0.{Map.containsKey}("hello")) && (let final dynamic #0 = world = #0#3 in true) && (!(#0#6 == null) || null is dynamic && #0#0.{Map.containsKey}("count")) && (let final dynamic #1 = count = #0#6 in true))) throw new StateError("Pattern matching error");
  }
})
```

It looks like the if statement in the transformed AST is doing a bunch of error checking that users aren't going to write tests for, so the whole destructuring expression is marked as not hit. The usual fix is to mark those nodes `isSynthetic` or just remove their `fileOffset`, but in this case it's hard to tell exactly where all the nodes are being created (the code around `ConstantsTransformer.visitPatternVariableDeclaration` is very abstracted, so it's hard to follow since I don't work on the CFE very often). I've tried removing the `fileOffset` for just about every node I can find around this area, but no luck.

I think we're going to need to add a bunch of plumbing to mark these nodes as synthetic. `MatchingExpressionVisitor` and `MatchingCache` look relevent. Any suggestions?

**Comment by johnniwinther** [comment]
I agree that we need extra plumbing in `MatchingExpressionVisitor` and `MatchingCache`. Maybe we also need a better way to signal this; `isSynthetic` is a bit too vague - leading to inconsistent interpretations in the various backends. We probably need to distinguish (at least) "don't show variable in debugger", "don't stop on this line in the debugger", and "don't include in coverage", and add the necessary flags to the AST.

(Part of) the property you're looking for in the case comes from whether the matching is irrefutable (which is why some of it was inadvertently fixed by my CL) since the checking in these cases is only for soundness.
